### PR TITLE
Jetty 9.4.x #2695 hpack compliance

### PR DESCRIPTION
**Pull request by gregw**
#2695 Hpack compliance

### CODE REVIEW

**Comment by gregw** [comment]
@sbordet this is my work in progress.  I have a few train trips coming up so I should make progress in the coming week

File: jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/HpackException.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    
+    public static class Stream extends HpackException
+    {
+        Stream(String messageFormat, Object... args)
+        {
+            super(messageFormat,args);
+        }
+    }
+    
+
```
Comment (sbordet) [inline jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/HpackException.java]: Remove unnecessary empty lines.

File: jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/HpackDecoder.java [hunk 2]
```diff
@@ -66,14 +66,14 @@ public void setLocalMaxDynamicTableSize(int localMaxdynamciTableSize)
...
-    public MetaData decode(ByteBuffer buffer)
+    public MetaData decode(ByteBuffer buffer) throws HpackException
     {
         if (LOG.isDebugEnabled())
             LOG.debug(String.format("CtxTbl[%x] decoding %d octets",_context.hashCode(),buffer.remaining()));
 
         // If the buffer is big, don't even think about decoding it
         if (buffer.remaining()>_builder.getMaxSize())
-            throw new BadMessageException(HttpStatus.REQUEST_HEADER_FIELDS_TOO_LARGE_431,"Header frame size "+buffer.remaining()+">"+_builder.getMaxSize());
+            throw new HpackException.Session("431 Request Header Fields too large",6);
```
Comment (sbordet) [inline jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/HpackDecoder.java]: What is the second parameter `6`?

File: jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/HpackException.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import java.io.IOException;
+
+public abstract class HpackException extends RuntimeException
+{
+    HpackException(String messageFormat, Object... args)
+    {
+        super(String.format(messageFormat, args));
+    }
+    
+    public static class Stream extends HpackException
```
Comment (sbordet) [inline jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/HpackException.java]: I think this class should be renamed `StreamException`. I understand there is duplication (as `HpackException.StreamException`, but it's immediately clear from the name that is an exception and not some data structure you can lookup from `HpackException`.
Also, the JDK always name exceptions ending with `*Exception` apart few historical cases.

File: jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/HpackException.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        Stream(String messageFormat, Object... args)
+        {
+            super(messageFormat,args);
+        }
+    }
+    
+    
+    public static class Session extends HpackException
```
Comment (sbordet) [inline jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/HpackException.java]: Ditto, `SessionException`.

File: jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/HpackException.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        Session(String messageFormat, Object... args)
+        {
+            super(messageFormat,args);
+        }
+    }
+    
+    public static class Compression extends Session
```
Comment (sbordet) [inline jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/HpackException.java]: Ditto, `CompressionException`.

File: jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/MetaDataBuilder.java [hunk 6]
```diff
@@ -134,29 +151,89 @@ else if (value != null)
...
+                case CONNECTION:
+                    // TODO should other connection specific fields be listed here?
+                    streamException("Connection specific field %s", header);
+                    break;                
 
-                default:
-                    if (name.charAt(0)!=':')
+                default:               
+                    if (name.charAt(0)==':')
+                        streamException("Unknown psuodo header %s", name);
```
Comment (sbordet) [inline jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/MetaDataBuilder.java]: `psuodo` -> `pseudo`

File: jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/MetaDataBuilder.java [hunk 7]
```diff
@@ -134,29 +151,89 @@ else if (value != null)
...
                 case CONTENT_LENGTH:
                     _contentLength = field.getLongValue();
                     _fields.add(field);
                     break;
+                
+                case TE:
+                    if ("trailors".equalsIgnoreCase(value))
+                        _fields.add(field);
+                    else
+                        streamException("unsupported TE value %s", value);
```
Comment (sbordet) [inline jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/MetaDataBuilder.java]: `unsupported` -> `Unsupported` as all other exception messages begins with upper case letter.

File: jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/MetaDataBuilder.java [hunk 8]
```diff
@@ -134,29 +151,89 @@ else if (value != null)
...
+                    _request = true;
                     break;
 
                 case CONTENT_LENGTH:
                     _contentLength = field.getLongValue();
                     _fields.add(field);
                     break;
+                
+                case TE:
+                    if ("trailors".equalsIgnoreCase(value))
```
Comment (sbordet) [inline jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/MetaDataBuilder.java]: "trailors"???

File: jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/MetaDataBuilder.java [hunk 9]
```diff
@@ -134,29 +151,89 @@ else if (value != null)
...
+                    else
                         _fields.add(field);
                     break;
             }
         }
         else
         {
-            if (name.charAt(0)!=':')
+            if (name.charAt(0)==':')
+                streamException("Unknown psuedo header %s",name);
```
Comment (sbordet) [inline jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/MetaDataBuilder.java]: `psuedo` -> `pseudo`

**Comment by sbordet** [comment]
@gregw I pushed branch `jetty-9.4.x-2679-h2spec_compliance` with the fixes for `h2spec`.

When you're happy with this branch, I will merge this branch into `jetty-9.4.x-2679-h2spec_compliance` as a single commit, add the new plumbing to catch the `HpackException`s and re-verify `h2spec` compliance.

**Pull request closed without merging**